src/design_docs/async_stack_traces.md

```diff
@@ -0,0 +1,705 @@
+            at C:\Users\ericholk\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.13.0\src\park\thread.rs:263
+18: tokio::runtime::enter::Enter::block_on<core::future::from_generator::GenFuture<async_tokio::main::generator$0> >
+            at C:\Users\ericholk\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.13.0\src\runtime\enter.rs:151
+19: tokio::runtime::thread_pool::ThreadPool::block_on<core::future::from_generator::GenFuture<async_tokio::main::generator$0> >
+            at C:\Users\ericholk\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.13.0\src\runtime\thread_pool\mod.rs:77
+20: tokio::runtime::Runtime::block_on<core::future::from_generator::GenFuture<async_tokio::main::generator$0> >
+            at C:\Users\ericholk\.cargo\registry\src\github.com-1ecc6299db9ec823\tokio-1.13.0\src\runtime\mod.rs:463
+21: async_tokio::main
+            at .\src\main.rs:4
+22: core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\ops\function.rs:227
+```
+
+If we remove these frames, we have something pretty close to the synchronous short backtrace:
+
+```
+   0: std::panicking::begin_panic_handler
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\std\src\panicking.rs:517
+   1: core::panicking::panic_fmt
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\panicking.rs:101
+   2: core::panicking::panic
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\/library\core\src\panicking.rs:50
+   3: common::baz::generator$0
+            at C:\Users\ericholk\repo\backtrace-examples\async-common\src\lib.rs:10
+   4: core::future::from_generator::impl$1::poll<common::baz::generator$0>
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\future\mod.rs:80
+   5: common::bar::generator$0
+            at C:\Users\ericholk\repo\backtrace-examples\async-common\src\lib.rs:6
+   6: core::future::from_generator::impl$1::poll<common::bar::generator$0>
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\future\mod.rs:80
+   7: common::foo::generator$0
+            at C:\Users\ericholk\repo\backtrace-examples\async-common\src\lib.rs:2
+   8: core::future::from_generator::impl$1::poll<common::foo::generator$0>
+            at /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35\library\core\src\future\mod.rs:80
+   9: async_tokio::main::generator$0
+            at .\src\main.rs:4
+```
+
+To achieve parity with the synchronous backtrace, there are two more improvements needed.
+
+The first is that in between each frame in user code, there is a frame with a call to `core::future::from_generator`.
+
+The second is that rather than seeing a call to `common::foo` or similar, we see `common::foo::generator$0`.
+
+These two are relatively minor issues.
+Fixing them may not be desirable.

```

Comment:
I disagree, I think that in aggreagate, these issues make async backtraces pretty painful and it is very desirable to fix them :-)

Reply:
Yeah.. I feel like when I wrote this I thought there was a good reason for doing it this way, but looking at it a month later I'm not sure why.

In the second case (`common::foo::generator$0`), maybe I thought that the user would care that there is a generator involved. On second thought, I don't know why they would. It's not like they wrote a generator, so having the one that rustc generator leak out this way is probably surprising and confusing.

On the first case, I think I was nervous about completely eliding stack frames from the user. Again, it's not clear why this information is useful, but I'd feel a little cheated to find out the debugger was hiding stack frames from me that were actually there. That said, because of inlining, the debugger is probably already telling me something that's not quite accurate.